src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt

```diff
@@ -123,41 +124,46 @@ private fun getDirectAndAllDependencies(
 }
 
 /**
- * Look for the prelude.
- * If the dependency defines a prelude, we overwrite an already defined prelude.
- * This is necessary to import the "std" prelude if a crate depends on both "core" and "std".
+ * Finds prelude based on edition and root module attributes such as ```#[no_std]```.
+ * Prelude for current crate can be overwritten by ```#[prelude_import]```.
+ * See also:
+ * - https://github.com/rust-lang/rust/blob/master/compiler/rustc_builtin_macros/src/standard_library_imports.rs
+ * - https://github.com/rust-lang/rust/pull/82217
  */
-private fun findPrelude(crate: Crate, allDependenciesDefMaps: Map<Crate, CrateDefMap>): ModData? {
-    /**
-     * Correct prelude is always selected (core vs std), because [Crate.flatDependencies] is top sorted.
-     * And we peek prelude from latest dependency.
-     */
-    val dependencies = crate.dependencies.mapToSet { it.crate }
-    val dependenciesTopSorted = crate.flatDependencies.filter { it in dependencies }
-    return dependenciesTopSorted
-        .asReversed()
-        .asSequence()
-        .mapNotNull { allDependenciesDefMaps[it]?.prelude }
-        .firstOrNull()
+private fun injectPrelude(defMap: CrateDefMap) {
+    val preludeCrate = when (defMap.stdlibAttributes) {
+        RsFile.Attributes.NONE -> AutoInjectedCrates.STD
+        RsFile.Attributes.NO_STD -> AutoInjectedCrates.CORE
+        RsFile.Attributes.NO_CORE -> return
+    }
+    val preludeName = when (val edition = defMap.metaData.edition) {
+        // We don't use "rust_2015" and "rust_2018" in order to be compatible with old rustc
```

Comment:
Let's add `// BACKCOMPAT: Rust %insert_necessary_version_here%` to mark this code as compatibility one